app/services/source-filters.ts

```diff
@@ -91,25 +112,84 @@ class SourceFiltersViews extends ViewHandler<{}> {
   }
 
   parsePresetValue(path: string) {
-    return path.match(/luts\\[a-z_]+.png$/)[0];
+    return path.match(/luts[\\\/][a-z_]+.png$/)[0];
+  }
+
+  filtersBySourceId(sourceId: string) {
+    return this.state.filters[sourceId];
+  }
+
+  presetFilterBySourceId(sourceId: string) {
+    return this.state.presets[sourceId];
+  }
+
+  getTypesForSource(sourceId: string): ISourceFilterType[] {
+    const source = this.getServiceViews(SourcesService).getSource(sourceId);
+    return this.state.types.filter(filterType => {
+      if (filterType.type === 'face_mask_filter') return false;
+
+      /* Audio filters can be applied to audio sources. */
+      if (source.audio && filterType.audio) {
+        return true;
+      }
+
+      /* We have either a video filter or source */
+      /* Can't apply asynchronous video filters to non-asynchronous video sources. */
+      if (!source.async && filterType.async) {
+        return false;
+      }
+
+      /* Video filters can be applied to video sources. */
+      if (source.video && filterType.video) {
+        return true;
+      }
+
+      return false;
+    });
+  }
+
+  suggestName(sourceId: string, filterName: string): string {
+    return namingHelpers.suggestName(
+      filterName,
+      (name: string) => !!this.state.filters[sourceId].find(f => f.name === name),
+    );
   }
 }
 
-export class SourceFiltersService extends Service {
+@InitAfter('SourcesService')
+export class SourceFiltersService extends StatefulService<IFiltersServiceState> {
   @Inject() private sourcesService: SourcesService;
   @Inject() private windowsService: WindowsService;
   @Inject() private usageStatisticsService: UsageStatisticsService;
 
+  static initialState: IFiltersServiceState = { filters: {}, presets: {}, types: [] };
+
   filterAdded = new Subject<ISourceFilterIdentifier>();
   filterRemoved = new Subject<ISourceFilterIdentifier>();
   filterUpdated = new Subject<ISourceFilterIdentifier>();
   filtersReordered = new Subject<void>();
 
+  init() {
+    this.sourcesService.sourceAdded.subscribe(s => {
+      if (!this.state.filters[s.sourceId]) this.SET_FILTERS(s.sourceId, []);
+    });
+
+    this.sourcesService.sourceRemoved.subscribe(s => {
+      this.SET_FILTERS(s.sourceId, null);
```

Comment:
why do we set null here but empty array above?

Reply:
It's a little weird.  It's a dictionary of active sources.  So `null` or `undefined` means the source doesn't exist.  Whereas `[]` is a source that exists but doesn't have any filters.  I'm actually doing this a little wrong through, because I should be deleting the source id from the dictionary entirely, instead of just setting it to `null`.

Reply:
Pushed a commit to address this.